shared/bundle_analysis/parsers/v3.py

```diff
@@ -0,0 +1,424 @@
+                            Asset.name == filename,
+                            Bundle.name == self.session.bundle.name,
+                        )
+                        .one()
+                    )
+                    imported_assets[filename] = asset
+                except (NoResultFound, MultipleResultsFound):
+                    log.info(
+                        f'Asset not found for dynamic import: "{filename}"',
+                        exc_info=True,
+                    )
+                    raise
+
+            dynamic_imports_list.extend(
+                [
+                    dict(chunk_id=chunk.id, asset_id=asset.id)
+                    for asset in imported_assets.values()
+                ]
+            )
+
+        return dynamic_imports_list
+
+    def _create_associations(self):
+        # associate chunks to assets
+        inserts = []
+        assets: list[Asset] = (
+            self.db_session.query(Asset)
+            .filter(
+                Asset.session_id == self.session.id,
+            )
+            .all()
+        )
+
+        asset_name_to_id = {asset.name: asset.id for asset in assets}
+
+        chunks: list[Chunk] = (
+            self.db_session.query(Chunk)
+            .filter(
+                Chunk.session_id == self.session.id,
+            )
+            .all()
+        )
+
+        chunk_unique_id_to_id = {chunk.unique_external_id: chunk.id for chunk in chunks}
+
+        modules = (
+            self.db_session.query(Module)
+            .filter(
+                Module.session_id == self.session.id,
+            )
+            .all()
+        )
+
+        for chunk in chunks:
+            chunk_id = chunk.id
+            asset_names = self.chunk_asset_names_index[chunk.unique_external_id]
+            inserts.extend(
+                [
+                    dict(asset_id=asset_name_to_id[asset_name], chunk_id=chunk_id)
+                    for asset_name in asset_names
+                ]
+            )
+        if inserts:
+            self.db_session.execute(assets_chunks.insert(), inserts)
+
+        # associate modules to chunks
+        # FIXME: this isn't quite right - need to sort out how non-JS assets reference chunks
+        inserts = []
+
+        modules: list[Module] = self.db_session.query(Module).filter(
+            Module.session_id == self.session.id,
+        )
+        for module in modules:
+            module_id = module.id
+            chunk_unique_external_ids = self.module_chunk_unique_external_ids_index[
+                module.name
+            ]
+
+            inserts.extend(
+                [
+                    dict(
+                        chunk_id=chunk_unique_id_to_id[unique_external_id],
+                        module_id=module_id,
+                    )
+                    for unique_external_id in chunk_unique_external_ids
+                ]
+            )
+        if inserts:
+            self.db_session.execute(chunks_modules.insert(), inserts)
```

Comment:
Do we need any transaction safety/rollback on the `assets_chunks.insert()` in above and this `chunks_modules.insert()` if one succeeds but the next fails?

Reply:
This entire `parse` should be already in a transaction, we don't actually call `session.commit` in here which writes everything into the filesystem, that happens on the caller of the parser.